src/librustc_codegen_llvm/intrinsic.rs

```diff
@@ -478,7 +478,10 @@ pub fn codegen_intrinsic_call(
                 "load" => {
                     let ty = substs.type_at(0);
                     if int_type_width_signed(ty, cx).is_some() {
-                        let align = cx.align_of(ty);
+                        // FIXME(eddyb) Isn't it UB to use `pref` instead of `abi` here?
+                        // However, 64-bit atomic loads on `i686-apple-darwin` appear to
+                        // require `___atomic_load` with ABI-alignment, so it's staying.
+                        let align = cx.layout_of(ty).aligns.pref;
```

Comment:
For atomic loads/stores, the alignment here must be at least the size of the pointee type. We should add an assertion for that.

Similarly, over-estimating alignment (i.e. if alloca is align 4, but load is align 8) is UB as well.

---

I think what we should do here is:

1. ICE here if the *abi* (I think this is what influenced by the `#[align(x)]` attribute, right?) alignment is smaller than the size of the type being loaded/stored;
2. Somehow propagate the abi alignment from the `Atomic*` struct to the integral type stored within it (?);
3. In some layer above this, add a proper error that would not allow using the atomic load/store (somehow propagating the abi alignment from the `Atomic*` struct to the integral type stored within it)intrinsics with underaligned types.

Reply:
Notably both of these are documented on the [load instruction in langref](https://llvm.org/docs/LangRef.html#id193).

Reply:
I just moved code around, it was in the builder methods before, this discussion should be moved to https://github.com/rust-lang/rust/pull/53514#issuecomment-419660238 (using `pref` alignment here is likely the wrong thing, it just happens to produce the right values)